service/history/historyEngine.go

```diff
@@ -2270,20 +2270,85 @@ type updateWorkflowAction struct {
 	createDecision bool
 }
 
-func (e *historyEngineImpl) updateWorkflowExecutionWithAction(
+type updateWorkflowActionFunc func(mutableState mutableState) (*updateWorkflowAction, error)
+
+func (e *historyEngineImpl) updateWorkflowExecutionWithActionAndContextReload(
 	ctx ctx.Context,
 	domainID string,
 	execution workflow.WorkflowExecution,
-	action func(builder mutableState) (*updateWorkflowAction, error),
+	action updateWorkflowActionFunc,
 ) (retError error) {
 
+	// for execution with runID, just update with action
+	if execution.GetRunId() != "" {
+		workflowContext := e.getWorkflowLockedContext(ctx, domainID, execution)
+		return e.updateWorkflowExecutionWithContextAndAction(workflowContext, action)
+	}
+
+	// for execution without runID,
+	// reload current run in case continueAsNew caused unexpected workflow already completed error
+	for attempt := 0; attempt < conditionalRetryCount; attempt++ {
```

Comment:
plz change this function to return the corresponding workflow for simplicity
plz also add another function in history engine which call this logic (getting the workflow) and do corresponding update

Reply:
updated